pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
+		err = r.restartProxy(cr)
+		if err != nil {
+			return errors.Wrap(err, "restart proxy")
+		}
+	}
+
+	if restartPXC {
+		err = r.restartPXC(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart pxc")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret) (bool, bool, error) {
+	var restartPXC, restartProxy bool
+	um, err := users.NewManager(cr.Name+"-pxc", string(internalSysSecretObj.Data["root"]))
+	if err != nil {
+		return restartPXC, restartProxy, errors.Wrap(err, "new users manager")
+	}
+
+	var sysUsers []users.SysUser
+
+	for name, pass := range sysUsersSecretObj.Data {
+		hosts := []string{}
+
+		if string(sysUsersSecretObj.Data[name]) == string(internalSysSecretObj.Data[name]) {
+			continue
+		}
+
+		switch name {
+		case "root":
+			restartProxy = true
+			hosts = []string{"localhost", "%"}
+		case "xtrabackup":
+			restartProxy = true
+			restartPXC = true
+			hosts = []string{"localhost"}
+		case "monitor":
+			restartProxy = true
+			hosts = []string{"10.%", "%"}
+		case "clustercheck":
+			restartProxy = true
+			restartPXC = true
+			hosts = []string{"localhost"}
+		case "proxyadmin":
+			restartProxy = true
+			continue
+
+		case "pmmserver":
+			if cr.Spec.PMM.Enabled {
+				restartProxy = true
+				restartPXC = true
+				continue
+			}
+		}
+		user := users.SysUser{
+			Name:  name,
+			Pass:  string(pass),
+			Hosts: hosts,
+		}
+		sysUsers = append(sysUsers, user)
+	}
+
+	if len(sysUsers) > 0 {
+		err = um.UpdateUsersPass(sysUsers)
+		if err != nil {
+			return restartPXC, restartProxy, errors.Wrap(err, "update sys users pass")
+		}
+	}
+
+	return restartPXC, restartProxy, nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) restartPXC(cr *api.PerconaXtraDBCluster, newSecretDataHash string) error {
+	sfsPXC := appsv1.StatefulSet{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Name + "-pxc",
+		},
+		&sfsPXC,
+	)
+
+	if len(sfsPXC.Annotations) == 0 {
+		sfsPXC.Annotations = make(map[string]string)
+	}
+	sfsPXC.Spec.Template.Annotations["last-applied-secret"] = newSecretDataHash
+
+	err = r.client.Update(context.TODO(), &sfsPXC)
+	if err != nil {
+		return errors.Wrap(err, "update pxc sfs last-applied annotation")
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) restartProxy(cr *api.PerconaXtraDBCluster) error {
+	pvcProxy := corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cr.Namespace,
+			Name:      "proxydata-" + cr.Name + "-proxysql-0",
+		},
+	}
+	err := r.client.Delete(context.TODO(), &pvcProxy)
+	if err != nil {
+		return errors.Wrap(err, "delete proxy pvc")
+	}
```

Comment:
wow, why we need to delete PVC?

Reply:
removed https://github.com/percona/percona-xtradb-cluster-operator/pull/418/commits/6c99594a7fcf4bdb02009113fe96749e10a465f9